File: glide-im-sdk/src/main/java/pro/glideim/sdk/entity/UserInfo.java
package pro.glideim.sdk.entity;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.TreeMap;

import pro.glideim.sdk.api.group.GroupInfoBean;
import pro.glideim.sdk.api.user.UserInfoBean;

public class UserInfo {

    private final TreeMap<String, IMContacts> contactsMap = new TreeMap<>();

    public final IMSessionList sessionList = new IMSessionList();

    public long uid;
    public String avatar = "";
    public String nickname = "-";
    public String token;
    public ContactsChangeListener contactsChangeListener;

    public List<IMContacts> updateContacts(List<UserInfoBean> userInfoBeans) {
        List<IMContacts> res = new ArrayList<>();
        for (UserInfoBean userInfoBean : userInfoBeans) {
            IMContacts c = contactsMap.get(1 + "_" + userInfoBean.getUid());
            if (c == null) {
                continue;
            }
            c.title = userInfoBean.getNickname();
            c.avatar = userInfoBean.getAvatar();
            c.id = userInfoBean.getUid();
            c.type = 1;
            res.add(c);
        }
        return res;
    }

    public List<IMContacts> updateContactsGroup(List<GroupInfoBean> groupInfoBeans) {
        List<IMContacts> res = new ArrayList<>();
        for (GroupInfoBean groupInfoBean : groupInfoBeans) {
            IMContacts c = contactsMap.get(2 + "_" + groupInfoBean.getGid());
            if (c == null) {
                continue;
            }
            c.title = groupInfoBean.getName();
            c.avatar = groupInfoBean.getAvatar();
            c.id = groupInfoBean.getGid();
            c.type = 2;
            res.add(c);
        }
        return res;
    }

    public void addContacts(List<IMContacts> contacts) {
        for (IMContacts c : contacts) {
            addContacts(c);
        }
    }

    public void addContacts(IMContacts c) {
        contactsMap.put(c.type + "_" + c.id, c);
    }

    public List<IMContacts> getContacts() {
        return new ArrayList<>(contactsMap.values());
    }

    public List<Long> getContactsGroup() {
        List<Long> g = new ArrayList<>();
        for (IMContacts contact : getContacts()) {
            if (contact.type == 2) {
                g.add(contact.id);
            }
        }
        return g;
    }
}

File: app/src/main/java/pro/glideim/db/UserInfo.kt
package pro.glideim.db

import androidx.room.Entity
import androidx.room.PrimaryKey
import pro.glideim.sdk.api.user.UserInfoBean

@Entity
data class UserInfo(
    @PrimaryKey
    val uid: Long,
    val avatar: String,
    val nickname: String,
    val account: String,
    val updateAt: Long
) {
    companion object {
        fun fromUserInfoBean(u: UserInfoBean): UserInfo {
            return UserInfo(
                u.uid,
                u.avatar,
                u.nickname,
                u.account,
                updateAt = System.currentTimeMillis()
            )
        }
    }

    fun toUserInfoBean(): UserInfoBean {
        return UserInfoBean().apply {
            uid = this@UserInfo.uid
            avatar = this@UserInfo.avatar
            account = this@UserInfo.account
            nickname = this@UserInfo.nickname
        }
    }
}